crates/libs/windows/src/core/factory_cache.rs

```diff
@@ -92,11 +91,12 @@ pub fn factory<C: RuntimeName, I: Interface>() -> Result<I> {
         while let Some(pos) = path.rfind('.') {
             path = &path[..pos];
 
-            let library = core::slice::from_raw_parts_mut(heap_alloc(path.len() + 5)? as *mut u8, path.len() + 5);
```

Comment:
This was potentially UB as `from_raw_parts_mut` requires the memory to be initialized which `heap_alloc` does not guarantee.  I moved to using a `Vec` because I wasn't sure there was a reason for us to directly allocate memory ourselves vs using the std lib.

Reply:
The memory provided by HeapAlloc is initialized, it's just not initialized to zero. So I don't think there's a real UB case here. But this is certainly nicer to look at, so no complaints.

Reply:
Suggestion: On a more serious note, I think we should create a separate function for this chunk of code and write a test against it. It is critical that this code iterate through a given set of components in the correct order.

Reply:
> The memory provided by HeapAlloc is initialized

Where do you see this? The documentation seems ambigious to me. When the `HEAP_ZERO_MEMORY` is passed "The allocated memory will be initialized to zero. Otherwise, the memory is not initialized to zero." ([link](https://docs.microsoft.com/en-us/windows/win32/api/heapapi/nf-heapapi-heapalloc#parameters)). One could read this as otherwise the memory is "initialized but just not zero" (which seems to be your understanding) or "not initialized at all" (which was my understanding). Is there any way to confirm that `HeapAlloc` *must* return initialized memory? 

>  I think we should create a separate function for this chunk of code and write a test against it

This is a good idea! 😊

Reply:
> One could read this as otherwise the memory is "initialized but just not zero" (which seems to be your understanding) or "not initialized at all" (which was my understanding). Is there any way to confirm that `HeapAlloc` _must_ return initialized memory?

HeapAlloc, on success, will return a block of memory. The values stored in that block are arguably irrelevant. I don't think `from_raw_parts_mut` should care about what values we're pointing to. I suspect the `properly initialized` segment in the bullet `data must point to len consecutive properly initialized values of type T` is just an error.

But I still prefer your change!

Reply:
> I suspect the properly initialized segment in the bullet `...` is just an error in Rust docs

This might not be relevant because we're talking about a slice of `u8` which can be any value, but in general this comment is very correct. If we were trying to build a `&mut [bool]` then random bytes would 100% be incorrect as Rust only allows `bool`s to be `0` or `1`. A byte that is not `0` or `1` representing a `bool` is UB and can cause miscompilations. 

For a good introduction in to why it's not correct for invalid data to represent a type even if that type is not read, [check out Ralf Jung's blog post on the topic](https://www.ralfj.de/blog/2020/07/15/unused-data.html).  

I know it feels pedantic, but this crate is foundational, and we should be *very* careful about memory safety.